app/lib/search/index_ducene.dart

```diff
@@ -0,0 +1,245 @@
+      IndexHolderDirectory indexDir;
+      if (envConfig.duceneDir != null) {
+        // Temporarily everything is in memory, because ducene is not using async
+        // IO, and that blocks appengine's health checks, killing the machine
+        // before it got any chance to setup.
+        // TODO: fix ducene to handle async FS properly
+        // _fsDir = new Directory(envConfig.duceneDir);
+        // indexDir = new FSIndexHolderDirectory(_fsDir);
+        indexDir = new RAMIndexHolderDirectory();
+      } else {
+        indexDir = new RAMIndexHolderDirectory();
+      }
+      _index = await DirectoryHolder.open(indexDir);
+    });
+  }
+
+  @override
+  bool get isReady => _lastUpdated != null;
+
+  @override
+  Future<int> indexSize() async {
+    if (_fsDir == null) return -1;
+    return _fsDir.list(recursive: true).asyncMap((FileSystemEntity fse) async {
+      if (fse is File) {
+        return fse.length();
+      }
+      return 0;
+    }).fold(0, (int sum, int v) => sum + v);
+  }
+
+  @override
+  Future add(PackageDocument doc) async {
+    await _init();
+    final Document duceneDoc = new Document()
+      ..append('id', doc.url)
+      ..append('package', doc.package, analyzer: _analyzer)
+      ..append('version', doc.version)
+      ..append('devVersion', doc.devVersion)
+      ..append(
+        'description',
+        _normalizeText(doc.description, 500),
+        analyzer: _analyzer,
+        stored: false,
+      )
+      ..append(
+          'originalDescription', _normalizeWhitespaces(doc.description, 500))
+      ..append('lastUpdated', doc.lastUpdated)
+      ..append(
+        'readme',
+        _normalizeText(doc.readme, 2000),
+        analyzer: _analyzer,
+        stored: false,
+      )
+      ..append('popularity', doc.popularity.toString());
+    if (doc.detectedTypes != null && doc.detectedTypes.isNotEmpty) {
+      duceneDoc.append('detectedTypes', doc.detectedTypes);
+    }
+    await _index.updateDocuments([duceneDoc]);
+  }
+
+  @override
+  Future addAll(Iterable<PackageDocument> documents) async {
+    // Adding documents one-by-one, will need to experiment with better batch
+    // size when async disk index is available.
+    for (PackageDocument doc in documents) {
+      await add(doc);
+    }
+  }
+
+  @override
+  Future contains(String url, String version, String devVersion) async {
+    final IndexSearcher searcher = _index.newRealTimeIndexSearcher();
+    final query = new BoolQuery(Op.and)
+      ..append('id', url)
+      ..append('version', version)
+      ..append('devVersion', devVersion);
+    final TopDocs topDocs = await searcher.search(query, 1);
+    return topDocs.scoreDocs.length == 1;
+  }
+
+  @override
+  Future removeUrl(String url) async {
+    await _init();
+    await _index.deleteDocuments(new BoolQuery()..append('id', url));
+  }
+
+  @override
+  Future merge() async {
+    await _init();
+    await _index.forceMerge();
+    _lastUpdated = new DateTime.now();
```

Comment:
When a service is storing datetimes or returning them to other services, we should always use UTC time.

Maybe it would also make sense to add an `assert(lastUpdated.isUtc)` in the serialization logic.